crates/polars-ops/src/frame/pivot/unpivot.rs

```diff
@@ -150,26 +150,48 @@ pub trait UnpivotDF: IntoDf {
         ids.as_single_chunk_par();
         drop(ids_);
 
-        let mut values = Vec::with_capacity(on.len());
         let columns = self_.get_columns();
 
-        for value_column_name in &on {
-            variable_col.extend_constant(len, Some(value_column_name.as_str()));
-            // ensure we go via the schema so we are O(1)
-            // self.column() is linear
-            // together with this loop that would make it O^2 over `on`
-            let (pos, _name, _dtype) = schema.try_get_full(value_column_name)?;
-            let col = &columns[pos];
-            let value_col = col.cast(&st).map_err(
-                |_| polars_err!(InvalidOperation: "'unpivot' not supported for dtype: {}", col.dtype()),
-            )?;
-            values.extend_from_slice(value_col.chunks())
-        }
-        let values_arr = concatenate_owned_unchecked(&values)?;
-        // SAFETY:
-        // The give dtype is correct
-        let values =
-            unsafe { Series::from_chunks_and_dtype_unchecked(value_name, vec![values_arr], &st) };
+        let values = match st {
+            #[cfg(feature = "dtype-categorical")]
+            DataType::Categorical(_, _) => {
```

Comment:
I am not really happy with this. Every `append` call will merge the rev-maps, so I suspect this will have quadratic performance.

What we should do instead is create a new Categorical altogether. Or check if all categoricals come from the same `Rev-map` and if so, we can take the default branch.